File: examples/sdk/sdk.js
/**
 * ROMA Execution Bridge v2.1.0 — JavaScript/Node.js SDK
 * Full billing flow: GPU-sec + tokens + spend-caps + 402 handling.
 */

const BASE_URL = "http://localhost:8900";
const API_KEY = "your-api-key-here";

class SpendCapExceeded extends Error {
  constructor(message, remaining) {
    super(message);
    this.name = "SpendCapExceeded";
    this.remaining = remaining;
  }
}

class ROMAClient {
  constructor(baseUrl = BASE_URL, apiKey = API_KEY) {
    this.base = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
  }

  async _request(method, path, body = null) {
    const url = `${this.base}${path}`;
    const opts = {
      method,
      headers: { "x-api-key": this.apiKey, "Content-Type": "application/json" },
    };
    if (body) opts.body = JSON.stringify(body);
    const resp = await fetch(url, opts);

    if (resp.status === 402) {
      const data = await resp.json();
      throw new SpendCapExceeded(
        data.detail || "spend cap exceeded",
        data.remaining_cap
      );
    }
    if (resp.status === 401) throw new Error("Invalid API key");
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return resp.json();
  }

  // ── Jobs ──────────────────────────────────────────────────────────
  async submitJob({ task, gpuRequired = false, gpuType = "any",
                     inputTokens = 0, outputTokens = 0, plan = "free",
                     priority = 5 } = {}) {
    return this._request("POST", "/submit", {
      task, gpu_required: gpuRequired, gpu_type: gpuType,
      input_tokens: inputTokens, output_tokens: outputTokens,
      plan, priority,
    });
  }

  async jobStatus(jobId) { return this._request("GET", `/status/${jobId}`); }
  async completeJob(jobId) { return this._request("POST", `/complete/${jobId}`); }
  async listJobs() { return this._request("GET", "/jobs"); }

  // ── Billing ──────────────────────────────────────────────────────
  async getUsage() { return this._request("GET", "/usage"); }

  async getBalance() {
    const usage = await this.getUsage();
    return usage.balance_usd || 0;
  }
}

// ═══════════════════════════════════════════════════════════════════════
//  Example: Full job lifecycle with billing
// ═══════════════════════════════════════════════════════════════════════

async function demoFullLifecycle() {
  const client = new ROMAClient(BASE_URL, "roma-demo-key-2026");

  // 1. Check balance before submitting
  let usage = await client.getUsage();
  console.log(`Balance: $${usage.balance_usd.toFixed(4)}  ` +
    `Plan: ${usage.plan}  Spend cap: $${usage.spend_cap_usd.toFixed(2)}`);

  // 2. Submit LLM job with tokens
  let job;
  try {
    job = await client.submitJob({
      task: "Summarize quarterly report (LLM inference)",
      gpuRequired: true,
      gpuType: "A100",
      inputTokens: 8000,
      outputTokens: 2000,
      plan: "pro",
    });
    console.log(`Job created: ${job.job_id}  ` +
      `Estimated cost: $${job.estimated_cost_usd}  ` +
      `Remaining cap: $${job.spend_cap_remaining}`);
  } catch (e) {
    if (e instanceof SpendCapExceeded) {
      console.error(`❌ SPEND CAP EXCEEDED: ${e.message}`);
      console.error("   Upgrade plan or wait for billing cycle reset.");
      process.exit(1);
    }
    throw e;
  }

  // 3. Poll until complete
  for (let i = 0; i < 5; i++) {
    const status = await client.jobStatus(job.job_id);
    console.log(`  Status: ${status.status}`);
    if (status.status === "completed") break;
    await new Promise(r => setTimeout(r, 2000));
  }

  // 4. Complete job — actual billing
  await client.completeJob(job.job_id);

  // 5. Final balance
  const usage2 = await client.getUsage();
  console.log(`Final balance: $${usage2.balance_usd.toFixed(4)}`);
  console.log(`Charged: $${(usage2.balance_usd - usage.balance_usd).toFixed(6)}`);
}

demoFullLifecycle().catch(console.error);

File: examples/sdk/sdk.py
#!/usr/bin/env python3
"""ROMA Execution Bridge v2.1.0 — Python SDK with billing (GPU-sec + tokens + spend-caps)."""

import os
import requests
import time
import sys

BASE_URL = "http://localhost:8900"
API_KEY = "your-api-key-here"


class ROMAClient:
    """Minimal ROMA client with billing-aware error handling."""

    def __init__(self, base_url: str = BASE_URL, api_key: str = API_KEY):
        self.base = base_url.rstrip("/")
        self.api_key = api_key
        self._headers = {"x-api-key": api_key, "Content-Type": "application/json"}

    def _request(self, method: str, path: str, **kwargs) -> dict:
        url = f"{self.base}{path}"
        resp = requests.request(method, url, headers=self._headers, **kwargs)
        if resp.status_code == 402:
            body = resp.json()
            raise SpendCapExceeded(
                body.get("detail", "spend cap exceeded"),
                remaining=body.get("remaining_cap"),
            )
        if resp.status_code == 401:
            raise PermissionError("Invalid API key")
        resp.raise_for_status()
        return resp.json()

    # ── Jobs ──────────────────────────────────────────────────────────

    def submit_job(
        self,
        task: str,
        gpu_required: bool = False,
        gpu_type: str = "any",
        input_tokens: int = 0,
        output_tokens: int = 0,
        plan: str = "free",
        priority: int = 5,
    ) -> dict:
        """Submit a GPU/LLM job with billing metadata."""
        payload = {
            "task": task,
            "gpu_required": gpu_required,
            "gpu_type": gpu_type,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "plan": plan,
            "priority": priority,
        }
        return self._request("POST", "/submit", json=payload)

    def job_status(self, job_id: str) -> dict:
        return self._request("GET", f"/status/{job_id}")

    def complete_job(self, job_id: str) -> dict:
        """Mark job complete — triggers actual billing recalculation."""
        return self._request("POST", f"/complete/{job_id}")

    def list_jobs(self) -> dict:
        return self._request("GET", "/jobs")

    # ── Billing ───────────────────────────────────────────────────────

    def get_usage(self) -> dict:
        """Get tenant usage: GPU-sec, tokens, cost, spend-cap status."""
        return self._request("GET", "/usage")

    def get_balance(self) -> float:
        """Convenience — just the current balance."""
        usage = self.get_usage()
        return usage.get("balance_usd", 0.0)


class SpendCapExceeded(Exception):
    """Raised when tenant exceeds plan spend-cap (HTTP 402)."""

    def __init__(self, message: str, remaining: float | None = None):
        super().__init__(message)
        self.remaining = remaining


# ═══════════════════════════════════════════════════════════════════════
#  Example: Full job lifecycle with billing
# ═══════════════════════════════════════════════════════════════════════


def demo_full_lifecycle():
    client = ROMAClient(api_key=os.environ.get("ROMA_API_KEY", "YOUR_API_KEY"))

    # 1. Check balance before submitting
    usage = client.get_usage()
    print(
        f"Balance: ${usage['balance_usd']:.4f}  "
        f"Plan: {usage['plan']}  "
        f"Spend cap: ${usage['spend_cap_usd']:.2f}  "
        f"GPU used: {usage['total_gpu_seconds']}s  "
        f"Tokens: {usage['total_input_tokens']} in / {usage['total_output_tokens']} out"
    )

    # 2. Submit LLM job with token tracking
    try:
        job = client.submit_job(
            task="Summarize quarterly report (LLM inference)",
            gpu_required=True,
            gpu_type="A100",
            input_tokens=8000,
            output_tokens=2000,
            plan="pro",
        )
        print(
            f"Job created: {job['job_id']}  "
            f"Estimated cost: ${job.get('estimated_cost_usd', 'N/A'):.6f}  "
            f"Remaining cap: ${job.get('spend_cap_remaining', 'N/A'):.6f}"
        )
    except SpendCapExceeded as e:
        print(f"❌ SPEND CAP EXCEEDED: {e}")
        print("   Upgrade plan or wait for billing cycle reset.")
        sys.exit(1)

    # 3. Poll status
    job_id = job["job_id"]
    for _ in range(5):
        status = client.job_status(job_id)
        print(f"  Status: {status['status']}")
        if status["status"] == "completed":
            break
        time.sleep(2)

    # 4. Complete job — triggers actual billing
    completed = client.complete_job(job_id)
    print(f"Completed: {completed}")

    # 5. Check balance after
    usage2 = client.get_usage()
    print(f"Final balance: ${usage2['balance_usd']:.4f}")
    print(f"Charged: ${usage2['balance_usd'] - usage['balance_usd']:.6f}")


if __name__ == "__main__":
    demo_full_lifecycle()
